board/main.c

```diff
@@ -218,26 +273,58 @@ void usb_cb_ep2_out(void *usbdata, int len, bool hardwired) {
   }
 }
 
+usb_asm_buffer ep3_buffer = {.ptr = 0, .tail_size = 0, .counter = 0};
+
 // send on CAN
 void usb_cb_ep3_out(void *usbdata, int len, bool hardwired) {
   UNUSED(hardwired);
-  int dpkt = 0;
-  uint32_t *d32 = (uint32_t *)usbdata;
-  for (dpkt = 0; dpkt < (len / 4); dpkt += 4) {
-    CAN_FIFOMailBox_TypeDef to_push;
-    to_push.RDHR = d32[dpkt + 3];
-    to_push.RDLR = d32[dpkt + 2];
-    to_push.RDTR = d32[dpkt + 1];
-    to_push.RIR = d32[dpkt];
-
-    uint8_t bus_number = (to_push.RDTR >> 4) & CAN_BUS_NUM_MASK;
-    can_send(&to_push, bus_number, false);
+  uint8_t *usbdata8 = (uint8_t *)usbdata;
+  // Got first packet from a stream, resetting buffer and counter
+  if (usbdata8[0] == 0) {
+    ep3_buffer.counter = 0;
+    ep3_buffer.ptr = 0;
+    ep3_buffer.tail_size = 0;
+  }
+  // Assembling can message with data from buffer
+  if (usbdata8[0] == ep3_buffer.counter) {
+    uint8_t pos = 1;
+    ep3_buffer.counter++;
+    if (ep3_buffer.ptr != 0) {
+      if (ep3_buffer.tail_size <= 63U) {
+        CANPacket_t to_push;
+        (void)memcpy(&ep3_buffer.data[ep3_buffer.ptr], &usbdata8[pos], ep3_buffer.tail_size);
```

Comment:
Since we use a lot of `memcpy` now, it might be worth it to merge (or write a new implementation) https://github.com/commaai/panda/pull/197/files.